# Анализ развития игровой индустрии с 2000 по 2013 год
Дата: 09.05.2026

## Общие сведения
### Цели и задачи проекта

<p><b>Цель проекта</b></p>
<p>Подготовить срез данных для анализа развития игровой индустрии с 2000 по 2013 год</p>

<p><b>Задачи проекта</b></p>
<p><ol>
    <li>Ознакомиться с данными</li>
    <li>Провести предобработку и очистить данные от ошибок</li>
    <li>Отфильтровать данные по нужному диапазону лет (2000 - 2013 год)</li>
    <li>Провести категоризацию данных</li>
    <li>Выделить топ-7 платформ по количеству игр</li>
</ol></p>

### Описание данных
<p>Данные <code>games_data.csv</code> содержат информацию о продажах игр разных жанров и платформ, а также пользовательские и экспертные оценки игр:</p>
<p><ul>
    <li><code>Name</code> — название игры;</li>
    <li><code>Platform</code> — название платформы;</li>
    <li><code>Year of Release</code> — год выпуска игры;</li>
    <li><code>Genre</code> — жанр игры;</li>
    <li><code>NA sales</code> — продажи в Северной Америке (в миллионах проданных копий);</li>
    <li><code>EU sales</code> — продажи в Европе (в миллионах проданных копий);</li>
    <li><code>JP sales</code> — продажи в Японии (в миллионах проданных копий);</li>
    <li><code>Other sales</code> — продажи в других странах (в миллионах проданных копий);</li>
    <li><code>Critic Score</code> — оценка критиков (от 0 до 100);</li>
    <li><code>User Score</code> — оценка пользователей (от 0 до 10);</li>
    <li><code>Rating</code> — рейтинг организации ESRB (англ. Entertainment Software Rating Board). Эта ассоциация определяет рейтинг компьютерных игр и присваивает им подходящую возрастную категорию.</li></ul></p>

### Содержимое проекта

<p><b>Основные шаги проекта</b></p>
<ol>
    <li>Загружаем данные и знакомимся с ними</li>
    <li>Проверяем ошибки в данных:
        <ul>
            <li>Приводим названия столбцов к единому стилю</li>
            <li>Нормализуем типы данных в столбцах</li>
            <li>Нормализуем типы данных в столбцах</li>
            <li>Приводим данные к нужным регистрам (где это будет полезно)</li>
        </ul>
    </li>
    <li>Проверяем пропуски в данных:
        <ul>
            <li>Смотрим на количество пропусков в столбцах и их долю от общего количества</li>
            <li>Восстанавливаем данных в пропусках, где это возможно</li>
            <li>Удалаяем пропуски, где данные нам не критичны для дальнейшего анализа</li>
        </ul>
    </li>
    <li>Проверяем дуюликаты в данных:
        <ul>
            <li>Выявляем неявные дубликаты или проверяем, что отработали с ними ранее</li>
            <li>Выявляем явные дубликаты и убираем их</li>
        </ul>
    </li>
    <li>Фильруем данные по нужным нам условиям</li>
    <li>Категоризируем данные и присваем нужные категории</li>
    <li>Делаем выводы</li>

## Загрузка данных и знакомство с ними


Загрузим необходимые бибилиотеки.

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('games_data.csv')

Познакомимся с данными в датафрейме.

In [3]:
df.head()

,Name,Platform,Year of Release,Genre,NA sales,EU sales,JP sales,Other sales,Critic Score,User Score,Rating
0,Wii Sports,Wii,2006.0,Sports,41.36,28.96,3.77,8.45,76.0,8,E
1,Super Mario Bros.,NES,1985.0,Platform,29.08,3.58,6.81,0.77,NaN,NaN,NaN
2,Mario Kart Wii,Wii,2008.0,Racing,15.68,12.76,3.79,3.29,82.0,8.3,E
3,Wii Sports Resort,Wii,2009.0,Sports,15.61,10.93,3.28,2.95,80.0,8,E
4,Pokemon Red/Pokemon Blue,GB,1996.0,Role-Playing,11.27,8.89,10.22,1.00,NaN,NaN,NaN


In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Name             16954 non-null  object 
 1   Platform         16956 non-null  object 
 2   Year of Release  16681 non-null  float64
 3   Genre            16954 non-null  object 
 4   NA sales         16956 non-null  float64
 5   EU sales         16956 non-null  object 
 6   JP sales         16956 non-null  object 
 7   Other sales      16956 non-null  float64
 8   Critic Score     8242 non-null   float64
 9   User Score       10152 non-null  object 
 10  Rating           10085 non-null  object 
dtypes: float64(4), object(7)
memory usage: 1.4+ MB


In [5]:
df.isna().sum()

Name                  2
Platform              0
Year of Release     275
Genre                 2
NA sales              0
EU sales              0
JP sales              0
Other sales           0
Critic Score       8714
User Score         6804
Rating             6871
dtype: int64

Сделаем вывод по данным.

<b>Вывод по данным</b>
<p>На основе предоставленных данных в таблице <code>games_data.csv</code> можно сделать следующие выводы:
<ol>
    <li>В таблице 16956 и 11 столбцов;</li>
    <li>Названия столбцов соответствуют описанию данных;</li>
    <li>В таблице представлены следующие типы данных:
        <ul>
            <li>7 столбцов с типом <code>object;</code></li>
            <li>4 столбца с типом <code>float64.</code></li>
        </ul>
    </li>
    <li>6 столбцов содержат пропуски в ячейках данных:
        <ul>
            <li><code>Name</code> - 2 пропуска;</li>
            <li><code>Year of Release</code> - 275 пропусков;</li>
            <li><code>Genre</code> - 2 пропуска;</li>
            <li><code>Critic Score</code> - 8714 пропусков;</li>
            <li><code>User Score</code> - 6804 пропуска;</li>
            <li><code>Rating</code> - 6871 пропуск.</li>
        </ul>
    </li>
</ol>

Нужно быть внимательными с данными в столбцах <code>EU sales</code>, <code>JP sales</code> и <code>User Score</code>, поскольку по <code>df.head()</code> видно, что есть значения ячеек типа <code>float64</code>, а по <code>df.info()</code> видно, что тип ячеек все-таки <code>object</code>.

---

## Проверка ошибок в данных и их предобработка


### Названия, или метки, столбцов датафрейма

Выведем названия всех столбцов датафрейма и проверим их стиль написания.

In [6]:
df.columns

Index(['Name', 'Platform', 'Year of Release', 'Genre', 'NA sales', 'EU sales',
       'JP sales', 'Other sales', 'Critic Score', 'User Score', 'Rating'],
      dtype='object')

Приведем все столбцы к стилю snake case. Названия сделаем в нижнем регистре, а пробелы заменим на подчеркивания.

In [7]:
df.columns = [column_name.lower().replace(' ', '_') for column_name in df.columns]

In [8]:
df.columns

Index(['name', 'platform', 'year_of_release', 'genre', 'na_sales', 'eu_sales',
       'jp_sales', 'other_sales', 'critic_score', 'user_score', 'rating'],
      dtype='object')

In [9]:
df.head(1)

,name,platform,year_of_release,genre,na_sales,eu_sales,jp_sales,other_sales,critic_score,user_score,rating
0,Wii Sports,Wii,2006.0,Sports,41.36,28.96,3.77,8.45,76.0,8,E


### Типы данных

<p>Проанализируем типы данных

<p>В колонках <code>Year of Release</code>, <code>NA sales</code>, <code>Other sales</code>, <code>Critic Score</code> к типу данных вопросов нет. А вот в остальных интересно проверить, что там находиться по типу данных, поскольку тип <code>object</code> может говорить о нескольких типах данных в столбце.</p>
<p>Поэтому, проверим, что там находится.</p>

In [10]:
object_types_columns = ['name', 'year_of_release', 'platform', 'genre', 'eu_sales', 'jp_sales', 'user_score', 'rating']

def check_unique_column_types(column_name):
    unique_types = df[f'{column_name}'].apply(type).unique().tolist()
    print(f'Используемые типы данных в столбце {column_name} - {unique_types}')

for element in object_types_columns:
    check_unique_column_types(element)

Используемые типы данных в столбце name - [<class 'str'>, <class 'float'>]
Используемые типы данных в столбце year_of_release - [<class 'float'>]
Используемые типы данных в столбце platform - [<class 'str'>]
Используемые типы данных в столбце genre - [<class 'str'>, <class 'float'>]
Используемые типы данных в столбце eu_sales - [<class 'str'>]
Используемые типы данных в столбце jp_sales - [<class 'str'>]
Используемые типы данных в столбце user_score - [<class 'str'>, <class 'float'>]
Используемые типы данных в столбце rating - [<class 'str'>, <class 'float'>]


<p>Видим, что нужно поработать со столбцами:
    <ul>
        <li><code>name</code> - тут предполагаю, что должен быть один тип данных <code>str</code>;</li>
        <li><code>year_of_release</code> - тут предполагаю, что должен быть один тип данных <code>datetime</code>, но <code>datetime</code> нам не очень подходит, то преобразуем в <code>Int64</code>;</li>
        <li><code>genre</code> - тут предполагаю, что должен быть один тип данных <code>str</code>;</li>
        <li><code>eu_sales</code> - тут предполагаю, что должен быть другой тип данных <code>float</code>, а не <code>str</code>;</li>
        <li><code>jp_sales</code> - тут предполагаю, что должен быть другой тип данных <code>float</code>,  а не <code>str</code>;</li>
        <li><code>user_score</code> - тут предполагаю, что должен быть один тип данных <code>float</code>;</li>
        <li><code>rating</code> - тут предполагаю, что должен быть один тип данных <code>str</code>.</li>
    </ul>
</p>
<p>К столбцу <code>platform</code> вопросов нет, тип данных <code>str</code> нас устраивает.

Но, перед тем, как работать по основному сценарию, хочу решить вопрос с 2мя пропусками в столбце name

In [11]:
df[df['name'].isna()]

,name,platform,year_of_release,genre,na_sales,eu_sales,jp_sales,other_sales,critic_score,user_score,rating
661,NaN,GEN,1993.0,NaN,1.78,0.53,0.0,0.08,NaN,NaN,NaN
14439,NaN,GEN,1993.0,NaN,0.00,0.0,0.03,0.00,NaN,NaN,NaN


Эти пропуски нам точно не будут полезны, поэтому удалю их

In [12]:
df = df.dropna(subset=['name'])
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 16954 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16954 non-null  object 
 1   platform         16954 non-null  object 
 2   year_of_release  16679 non-null  float64
 3   genre            16954 non-null  object 
 4   na_sales         16954 non-null  float64
 5   eu_sales         16954 non-null  object 
 6   jp_sales         16954 non-null  object 
 7   other_sales      16954 non-null  float64
 8   critic_score     8242 non-null   float64
 9   user_score       10152 non-null  object 
 10  rating           10085 non-null  object 
dtypes: float64(4), object(7)
memory usage: 1.6+ MB


In [13]:
check_unique_column_types('name')

Используемые типы данных в столбце name - [<class 'str'>]


Заметим, что в столбце <code>genre</code> больше нет нулевых значений

In [14]:
check_unique_column_types('genre')

Используемые типы данных в столбце genre - [<class 'str'>]


Теперь поработаем с присвоением нужного типа данных, учитывая комментарий ниже, что могу встречаться строковые значения в числовых столбцах. Также стоит не забыть наши изменения выше, т.е. нам нужно теперь только поработать со столбцами <code>eu_sales</code>, <code>jp_sales</code>, <code>user_score</code> и <code>rating</code>.

Перед переводом к нужным типам интересно посмотреть, что за данные вообще могут быть в столбце, отличные от float или object

In [15]:
columns_to_float = ['eu_sales', 'jp_sales', 'user_score']

def check_float_type(column_name):
    float_values = set()
    for e in df[column_name]:
        if isinstance(e, float):
            float_values.add(e)
    print(f'Значения с типом float в {column_name} - {float_values}')

def check_numeric(column_name):
    non_numeric = set()
    for e in df[column_name]:
        try:
            float(e)
        except ValueError:
            non_numeric.add(e)
    print(f'Не конвертируемые значения {column_name} - {non_numeric}')  

for element in columns_to_float:
    check_numeric(element)
    
check_float_type('rating')

Не конвертируемые значения eu_sales - {'unknown'}
Не конвертируемые значения jp_sales - {'unknown'}
Не конвертируемые значения user_score - {'tbd'}
Значения с типом float в rating - {nan}


Видим, что ничего критичного в "ненужных" типах данных нет и можно смело менять типы данных в столбцах

Переведем столбцы `eu_sales`,`jp_sales` и `user_score` к типу `float`

In [16]:
df[['eu_sales', 'jp_sales', 'user_score']] = df[['eu_sales', 'jp_sales', 'user_score']].apply(pd.to_numeric, errors='coerce')

In [17]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 16954 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16954 non-null  object 
 1   platform         16954 non-null  object 
 2   year_of_release  16679 non-null  float64
 3   genre            16954 non-null  object 
 4   na_sales         16954 non-null  float64
 5   eu_sales         16948 non-null  float64
 6   jp_sales         16950 non-null  float64
 7   other_sales      16954 non-null  float64
 8   critic_score     8242 non-null   float64
 9   user_score       7688 non-null   float64
 10  rating           10085 non-null  object 
dtypes: float64(7), object(4)
memory usage: 1.6+ MB


Работы с преобразованием типов мы закончили. Но для удобства работы с данными приведем значения столбцов к нужным регистрам:
<ul>
    <li><code>name</code> - приведем к нижнему регистру;</li>
    <li><code>platform</code> - приведем к высшему регистру;</li>
    <li><code>genre</code> - приведем к нижнему регистру;</li>
    <li><code>rating</code> - приведем к высшему регистру.</li>
</ul>

In [18]:
df['name'] = df['name'].str.lower()
df['platform'] = df['platform'].str.upper()
df['genre'] = df['genre'].str.lower()
df['rating'] = df['rating'].str.upper()
df.head(10)

,name,platform,year_of_release,genre,na_sales,eu_sales,jp_sales,other_sales,critic_score,user_score,rating
0,wii sports,WII,2006.0,sports,41.36,28.96,3.77,8.45,76.0,8.0,E
1,super mario bros.,NES,1985.0,platform,29.08,3.58,6.81,0.77,NaN,NaN,NaN
2,mario kart wii,WII,2008.0,racing,15.68,12.76,3.79,3.29,82.0,8.3,E
3,wii sports resort,WII,2009.0,sports,15.61,10.93,3.28,2.95,80.0,8.0,E
4,pokemon red/pokemon blue,GB,1996.0,role-playing,11.27,8.89,10.22,1.00,NaN,NaN,NaN
5,tetris,GB,1989.0,puzzle,23.20,2.26,4.22,0.58,NaN,NaN,NaN
6,new super mario bros.,DS,2006.0,platform,11.28,9.14,6.50,2.88,89.0,8.5,E
7,wii play,WII,2006.0,misc,13.96,9.18,2.93,2.84,58.0,6.6,E
8,new super mario bros. wii,WII,2009.0,platform,14.44,6.94,4.70,2.24,87.0,8.4,E
9,duck hunt,NES,1984.0,shooter,26.93,0.63,0.28,0.47,NaN,NaN,NaN


### Наличие пропусков в данных

Посчитаем количество пропусков в каждом столбце в абсолютных и относительных значениях.

In [19]:
df.isna().sum()

name                  0
platform              0
year_of_release     275
genre                 0
na_sales              0
eu_sales              6
jp_sales              4
other_sales           0
critic_score       8712
user_score         9266
rating             6869
dtype: int64

In [20]:
df_isna_ratio = df.isna().mean() * 100
df_isna_ratio

name                0.000000
platform            0.000000
year_of_release     1.622036
genre               0.000000
na_sales            0.000000
eu_sales            0.035390
jp_sales            0.023593
other_sales         0.000000
critic_score       51.386104
user_score         54.653769
rating             40.515513
dtype: float64

In [21]:
df_isna_ratio.sort_values(ascending = False)

user_score         54.653769
critic_score       51.386104
rating             40.515513
year_of_release     1.622036
eu_sales            0.035390
jp_sales            0.023593
name                0.000000
platform            0.000000
genre               0.000000
na_sales            0.000000
other_sales         0.000000
dtype: float64

In [22]:
df[df['year_of_release'].isna()]

,name,platform,year_of_release,genre,na_sales,eu_sales,jp_sales,other_sales,critic_score,user_score,rating
183,madden nfl 2004,PS2,NaN,sports,4.26,0.26,0.01,0.71,94.0,8.5,E
379,fifa soccer 2004,PS2,NaN,sports,0.59,2.36,0.04,0.51,84.0,6.4,E
458,lego batman: the videogame,WII,NaN,action,1.80,0.97,0.00,0.29,74.0,7.9,E10+
477,wwe smackdown vs. raw 2006,PS2,NaN,fighting,1.57,1.02,0.00,0.41,NaN,NaN,NaN
611,space invaders,2600,NaN,shooter,2.36,0.14,0.00,0.03,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...
16609,pdc world championship darts 2008,PSP,NaN,sports,0.01,0.00,0.00,0.00,43.0,NaN,E10+
16641,freaky flyers,GC,NaN,racing,0.01,0.00,0.00,0.00,69.0,6.5,T
16685,inversion,PC,NaN,shooter,0.01,0.00,0.00,0.00,59.0,6.7,M
16695,hakuouki: shinsengumi kitan,PS3,NaN,adventure,0.01,0.00,0.00,0.00,NaN,NaN,NaN


Изучим данные с пропущенными значениями.

<p>У нас основные столбцы с пропусками:</p>
<ul>
    <li>year_of_release;</li>
    <li>eu_sales;</li>
    <li>jp_sales;</li>
    <li>critic_score;</li>
    <li>user_score;</li>
    <li>rating.</li>
</ul>
<p>Из них для нашей задачи важными являются:</p>
<ul>
    <li>year_of_release;</li>
    <li>critic_score;</li>
    <li>user_score.</li>
</ul>
<p>Попробуем разобраться с каждым</p>
<p><b>year_of_release</b></p>
<ol>
    <li>Доля пропусков составляет всего 1.5%, что не так много;</li>
    <li>В некоторых названиях присутствует год (например, Madden NFL 2004 или FIFA Soccer 2004). В теории можно попробовать восстановить год из названия, поскольку даже в примерах игры попадают в нужный нам временной диапазон. Но сделаем по-другому, поскольку пропусков по названиям не так много, всего 275, мы спокойно можем поискать информацию по ним в сети и точно восстановить год.</li>
</ol>
<p><b>critic_score</b> и <b>user_score</b></p>
<ol>
    <li>Доля пропусков составляет 50% и 54% - это очень много;</li>
    <li>Есть данные, где <code>critic_score</code> и <code>user_score</code> могут быть одновременно быть NaN, а где по отдельности;</li>
    <li>Предалагается сначала разобраться с годами, а потом смотреть на эти показатели.</li>
</ol>

Поработаем с <code>year_of_release</code>. Создадим отдельный датафрейм со столбцами <code>name</code> и пустыми <code>year_of_release</code>.

In [23]:
df_year_nan = df.loc[df['year_of_release'].isna(), ['name', 'platform', 'year_of_release']]
df_year_nan.info()

<class 'pandas.core.frame.DataFrame'>
Index: 275 entries, 183 to 16760
Data columns (total 3 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             275 non-null    object 
 1   platform         275 non-null    object 
 2   year_of_release  0 non-null      float64
dtypes: float64(1), object(2)
memory usage: 8.6+ KB


In [24]:
df_year_nan.head(10)

,name,platform,year_of_release
183,madden nfl 2004,PS2,NaN
379,fifa soccer 2004,PS2,NaN
458,lego batman: the videogame,WII,NaN
477,wwe smackdown vs. raw 2006,PS2,NaN
611,space invaders,2600,NaN
629,rock band,X360,NaN
659,frogger's adventures: temple of the frog,GBA,NaN
680,lego indiana jones: the original adventures,WII,NaN
722,call of duty 3,WII,NaN
808,rock band,WII,NaN


Соберем список всех наименований

In [25]:
missing_year_list = df_year_nan['name'].unique().tolist()
missing_year_list

['madden nfl 2004',
 'fifa soccer 2004',
 'lego batman: the videogame',
 'wwe smackdown vs. raw 2006',
 'space invaders',
 'rock band',
 "frogger's adventures: temple of the frog",
 'lego indiana jones: the original adventures',
 'call of duty 3',
 'call of duty: black ops',
 'triple play 99',
 'adventure',
 'combat',
 'lego harry potter: years 5-7',
 'nascar thunder 2003',
 'hitman 2: silent assassin',
 'legacy of kain: soul reaver',
 'donkey kong land iii',
 'air-sea battle',
 'suikoden iii',
 'wheel of fortune',
 'yakuza 4',
 'namco museum',
 'rhythm heaven',
 'the lord of the rings: war in the north',
 'madden nfl 07',
 'mlb slugfest 20-03',
 'shaun white snowboarding',
 'pes 2009: pro evolution soccer',
 'warioware: twisted!',
 'madden nfl 11',
 'test drive unlimited 2',
 'the chronicles of narnia: the lion, the witch and the wardrobe',
 'monster hunter 2',
 'metal gear solid 2: substance',
 'advance wars: days of ruin',
 'the golden compass',
 'madden nfl 06',
 'nascar: dirt to d

In [26]:
len(missing_year_list)

232

Применив метод <code>unique()</code> мы для себя сразу отметили наличие дубликатов в названии. Это нам на будущее, что обязательно надо будет поработать с дубликатами. а пока поищем в сети данные по играм и составим словарь, где ключ - наименование, а значение - год.

In [27]:
games_release_years = {
    'madden nfl 2004': 2003,
    'fifa soccer 2004': 2003,
    'lego batman: the videogame': 2008,
    'wwe smackdown vs. raw 2006': 2005,
    'space invaders': 1978,
    'rock band': 2007,
    "frogger's adventures: temple of the frog": 2001,
    'lego indiana jones: the original adventures': 2008,
    'call of duty 3': 2006,
    'call of duty: black ops': 2010,
    'triple play 99': 1998,
    'adventure': 1980,
    'combat': 1977,
    'lego harry potter: years 5-7': 2011,
    'nascar thunder 2003': 2002,
    'hitman 2: silent assassin': 2002,
    'legacy of kain: soul reaver': 1999,
    'donkey kong land iii': 1997,
    'air-sea battle': 1977,
    'suikoden iii': 2002,
    'wheel of fortune': 1987,
    'yakuza 4': 2010,
    'namco museum': 1995,
    'rhythm heaven': 2008,
    'the lord of the rings: war in the north': 2011,
    'madden nfl 07': 2006,
    'mlb slugfest 20-03': 2002,
    'shaun white snowboarding': 2008,
    'pes 2009: pro evolution soccer': 2008,
    'warioware: twisted!': 2004,
    'madden nfl 11': 2010,
    'test drive unlimited 2': 2011,
    'the chronicles of narnia: the lion, the witch and the wardrobe': 2005,
    'monster hunter 2': 2004,
    'metal gear solid 2: substance': 2002,
    'advance wars: days of ruin': 2008,
    'the golden compass': 2007,
    'madden nfl 06': 2005,
    'nascar: dirt to daytona': 2004,
    'madden nfl 2002': 2001,
    'def jam: fight for ny': 2004,
    'nba street vol. 2': 2003,
    'fishing derby': 1980,
    'wet': 2009,
    'sonic the hedgehog': 1991,
    'karate': 1986,
    'tiger woods pga tour 07': 2006,
    'circus atari': 1982,
    'the chronicles of riddick: escape from butcher bay': 2004,
    "maze craze: a game of cops 'n robbers": 1978,
    'silent hill: homecoming': 2008,
    'super breakout': 1978,
    "robert ludlum's the bourne conspiracy": 2008,
    'nhl slapshot': 2010,
    'tera': 2011,
    'nfl gameday 2003': 2002,
    'harvest moon: save the homeland': 2001,
    'hangman': 1978,
    'nba live 2003': 2002,
    "cubix robots for everyone: clash 'n' bash": 2003,
    'dragon ball z: budokai tenkaichi 2 (jp sales)': 2006,
    'tropico 4': 2011,
    'tomb raider (2013)': 2013,
    'bejeweled 3': 2010,
    'custom robo': 2004,
    'final fantasy xi': 2002,
    'singularity': 2010,
    'dragster': 1980,
    'all-star baseball 2005': 2004,
    'star wars jedi knight ii: jedi outcast': 2002,
    'slot machine': 1979,
    'the dukes of hazzard ii: daisy dukes it out': 2000,
    'harvest moon: the tale of two towns': 2011,
    'shrek the third': 2007,
    'nicktoons: battle for volcano island': 2006,
    'haven: call of the king': 2002,
    'unreal championship 2: the liandri conflict': 2003,
    'pac-man fever': 2002,
    'the legend of zelda: the minish cap(weekly jp sales)': 2004,
    'indy 500': 1977,
    'disgaea 3: absence of detention': 2008,
    'flag capture': 1978,
    'gun': 2005,
    'rock revolution': 2008,
    'college hoops 2k6': 2005,
    'jonah lomu rugby challenge': 2011,
    'mega man x collection': 2006,
    'bioshock 2': 2010,
    'danganronpa: trigger happy havoc': 2010,
    'jet x20': 2002,
    "tony hawk's downhill jam": 2006,
    'tribes: aerial assault': 2002,
    'big beach sports 2': 2010,
    'yu yu hakusho: dark tournament': 2004,
    'ghostbusters ii': 1989,
    'breakaway iv': 1997,
    'robotech: battlecry': 2002,
    'move fitness': 2011,
    'valkyria chronicles iii: unrecorded chronicles': 2011,
    'dancedancerevolution ii': 1999,
    'wrc: fia world rally championship': 2010,
    'famista 64': 1997,
    'dead space 3': 2013,
    'pet zombies': 2011,
    'star trek: legacy': 2006,
    'backbreaker': 2010,
    'twisted metal: small brawl': 2001,
    'otomedius excellent': 2011,
    'nba starting five': 2002,
    'teen titans': 2006,
    'trauma team': 2010,
    "james cameron's dark angel": 2002,
    'sword of the samurai': 1989,
    'splatterhouse': 1988,
    'alone in the dark: the new nightmare': 2001,
    'vegas party': 2010,
    'jurassic park: the game': 2011,
    'home run': 1977,
    'ejay clubworld': 2003,
    'our house party!': 2011,
    'wcw backstage assault': 2000,
    "disney's cinderella: magical dreams": 2005,
    'transworld surf': 2002,
    'street fighter iv': 2008,
    'nintendo puzzle collection': 2003,
    'charm girls club: my fashion mall': 2010,
    'record of agarest war zero': 2009,
    'rocksmith': 2011,
    'super robot wars og saga: masou kishin ii - revelation of evil god': 2012,
    'saru! get you! million monkeys': 2006,
    'street hoops': 2000,
    'godzilla: destroy all monsters melee': 2002,
    'the daring game for girls': 2008,
    'major league baseball 2k6': 2006,
    'star trek: conquest': 2006,
    'giftpia': 2003,
    'happy feet two': 2011,
    "disney's chicken little: ace in action": 2005,
    'atsumare! power pro kun no ds koushien': 2007,
    'my healthy cooking coach': 2008,
    'luminous arc 2 (jp sales)': 2008,
    'egg monster hero': 2004,
    'demon chaos': 2001,
    'samurai shodown anthology': 2008,
    'action man-operation extreme': 2003,
    'super puzzle fighter ii': 1996,
    'charm girls club: my fashion show': 2010,
    'face racers: photo finish': 2011,
    'zero: tsukihami no kamen': 2008,
    'the hidden': 2011,
    'get fit with mel b': 2011,
    'mega man battle network: operation shooting star': 2008,
    'smashing drive': 2002,
    'port royale 3': 2012,
    'dream trigger 3d': 2011,
    'dead island: riptide': 2013,
    'yoostar on mtv': 2011,
    'tornado': 1988,
    "mcfarlane's evil prophecy": 2004,
    'drake of the 99 dragons': 2003,
    'build-a-bear workshop: friendship valley': 2010,
    'rayman arena': 2003,
    'national geographic challenge!': 2010,
    'alex rider: stormbreaker': 2006,
    'chou soujuu mecha mg': 1999,
    'prinny: can i really be the hero? (us sales)': 2009,
    'combat elite: wwii paratroopers': 2005,
    "flip's twisted world": 2010,
    'mobile ops: the one year war': 2009,
    "tom clancy's rainbow six: critical hour": 2006,
    'jewel link chronicles: mountains of madness': 2011,
    'captain america: super soldier': 2011,
    'mountain bike adrenaline': 2009,
    'drill dozer': 2005,
    'grid': 2008,
    'tour de france 2011': 2011,
    'reader rabbit 2nd grade': 2000,
    'monster hunter frontier online': 2007,
    'rollercoaster tycoon': 1999,
    'battle vs. chess': 2011,
    'the history channel: great battles - medieval': 2010,
    'clockwork empires': 2015,
    'b.l.u.e.: legend of water': 2008,
    'nhl hitz pro': 2003,
    "luxor: pharaoh's challenge": 2009,
    'sega rally 2006': 2005,
    'half-minute hero 2': 2011,
    'housekeeping': 2011,
    'major league baseball 2k8': 2008,
    'sabre wulf': 1984,
    'swords': 2011,
    'beyond the labyrinth': 2009,
    'bikkuriman daijiten': 1988,
    'majesty 2: the fantasy kingdom sim': 2009,
    'fullmetal alchemist: brotherhood': 2009,
    'samurai spirits: tenkaichi kenkakuden': 2008,
    'world of tanks': 2010,
    'tom and jerry in war of the whiskers': 2002,
    'super duper sumos': 2001,
    'the king of fighters: maximum impact - maniax': 2004,
    'combat wings: the great battles of wwii': 2011,
    'tube slider': 2003,
    'umineko no naku koro ni san: shinjitsu to gensou no yasoukyoku': 2010,
    'payout poker & casino': 2009,
    'wii de asobu: metroid prime': 2009,
    'legacy of ys: books i & ii': 2009,
    'saint': 2009,
    'steal princess': 2008,
    'mario tennis': 2000,
    'runaway: a twist of fate': 2006,
    "yu-gi-oh! 5d's wheelie breakers (jp sales)": 2009,
    "cabela's alaskan adventure": 2006,
    'writing and speaking beautiful japanese ds': 2008,
    'virtua quest': 2004,
    "shonen jump's yu-gi-oh! gx card almanac": 2005,
    'without warning': 2005,
    'pdc world championship darts 2008': 2007,
    'dinotopia: the sunstone odyssey': 2002,
    'jet impulse': 2006,
    'dream dancer': 2009,
    "dance! it's your stage": 2008,
    'football manager 2007': 2006,
    'ferrari: the race experience': 2010,
    'aquaman: battle for atlantis': 2003,
    'homeworld remastered collection': 2015,
    'shorts': 2009,
    'akb1/48: idol to guam de koishitara...': 2011,
    'brothers in arms: furious 4': 2012,
    'agarest senki: re-appearance': 2010,
    'freaky flyers': 2003,
    'inversion': 2012,
    'hakuouki: shinsengumi kitan': 2008
}

Данные нашли, теперь заполним датафрейм <code>df_year_nan</code> этими данными.

In [28]:
df_year_nan['year_of_release'] = df_year_nan['year_of_release'].fillna(df_year_nan['name'].map(games_release_years))
df_year_nan.head(10)

,name,platform,year_of_release
183,madden nfl 2004,PS2,2003.0
379,fifa soccer 2004,PS2,2003.0
458,lego batman: the videogame,WII,2008.0
477,wwe smackdown vs. raw 2006,PS2,2005.0
611,space invaders,2600,1978.0
629,rock band,X360,2007.0
659,frogger's adventures: temple of the frog,GBA,2001.0
680,lego indiana jones: the original adventures,WII,2008.0
722,call of duty 3,WII,2006.0
808,rock band,WII,2007.0


In [29]:
df_year_nan.info()

<class 'pandas.core.frame.DataFrame'>
Index: 275 entries, 183 to 16760
Data columns (total 3 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             275 non-null    object 
 1   platform         275 non-null    object 
 2   year_of_release  275 non-null    float64
dtypes: float64(1), object(2)
memory usage: 8.6+ KB


In [30]:
df_year_nan['year_of_release'].isna().sum()

np.int64(0)

Данные подставили, пропусков нет. Теперь перенесем данные в основной датафрем. Можем использовать проверку по индексу, поскольку они у нас совпадают.

In [31]:
df.loc[df_year_nan.index, 'year_of_release'] = df_year_nan['year_of_release']
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 16954 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16954 non-null  object 
 1   platform         16954 non-null  object 
 2   year_of_release  16954 non-null  float64
 3   genre            16954 non-null  object 
 4   na_sales         16954 non-null  float64
 5   eu_sales         16948 non-null  float64
 6   jp_sales         16950 non-null  float64
 7   other_sales      16954 non-null  float64
 8   critic_score     8242 non-null   float64
 9   user_score       7688 non-null   float64
 10  rating           10085 non-null  object 
dtypes: float64(7), object(4)
memory usage: 2.1+ MB


In [32]:
df.head(10)

,name,platform,year_of_release,genre,na_sales,eu_sales,jp_sales,other_sales,critic_score,user_score,rating
0,wii sports,WII,2006.0,sports,41.36,28.96,3.77,8.45,76.0,8.0,E
1,super mario bros.,NES,1985.0,platform,29.08,3.58,6.81,0.77,NaN,NaN,NaN
2,mario kart wii,WII,2008.0,racing,15.68,12.76,3.79,3.29,82.0,8.3,E
3,wii sports resort,WII,2009.0,sports,15.61,10.93,3.28,2.95,80.0,8.0,E
4,pokemon red/pokemon blue,GB,1996.0,role-playing,11.27,8.89,10.22,1.00,NaN,NaN,NaN
5,tetris,GB,1989.0,puzzle,23.20,2.26,4.22,0.58,NaN,NaN,NaN
6,new super mario bros.,DS,2006.0,platform,11.28,9.14,6.50,2.88,89.0,8.5,E
7,wii play,WII,2006.0,misc,13.96,9.18,2.93,2.84,58.0,6.6,E
8,new super mario bros. wii,WII,2009.0,platform,14.44,6.94,4.70,2.24,87.0,8.4,E
9,duck hunt,NES,1984.0,shooter,26.93,0.63,0.28,0.47,NaN,NaN,NaN


In [33]:
df['year_of_release'].isna().sum()

np.int64(0)

С годами разобрались, восстановили важную часть данных. Теперь нужно разобраться с пропусками критиков <code>critic_score</code> и пользователей <code>user_score</code>.
Мы знаем, что данные могут быть пропущегы и там, и там одновременно или по отдельности. Есть идея, проверить отношение рейтингов и попробовать на этом восстановить данные, где пропущено по одному значению.

In [34]:
df_with_rating = df.loc[(df['critic_score'].notna()) & (df['user_score'].notna()), ['name', 'critic_score', 'user_score']]
df_with_rating

,name,critic_score,user_score
0,wii sports,76.0,8.0
2,mario kart wii,82.0,8.3
3,wii sports resort,80.0,8.0
6,new super mario bros.,89.0,8.5
7,wii play,58.0,6.6
...,...,...,...
16913,mortal kombat: deadly alliance,81.0,8.8
16932,metal gear solid v: ground zeroes,80.0,7.6
16936,breach,61.0,5.8
16943,storm: frontline nation,60.0,7.2


Попробуем вычислить средний коэффициент отношения <code>critic_score</code> и <code>user_score</code>

In [35]:
ratio = df_with_rating['critic_score'] / df_with_rating['user_score']
ratio

0         9.500000
2         9.879518
3        10.000000
6        10.470588
7         8.787879
           ...    
16913     9.204545
16932    10.526316
16936    10.517241
16943     8.333333
16946    10.862069
Length: 7105, dtype: float64

In [36]:
ratio_mean = ratio.mean()
ratio_mean

np.float64(10.10655105748862)

Коэффициент мы вычилили, теперь для <code>critic_score</code> и <code>user_score</code> восстановим данные, но через промежуточные фреймы, чтобы себя перепроверить.

In [37]:
critic_mask = df.loc[(df['critic_score'].notna()) & (df['user_score'].isna()), ['name', 'critic_score', 'user_score']]
critic_mask

,name,critic_score,user_score
182,tomb raider ii,85.0,NaN
302,namco museum: 50th anniversary,61.0,NaN
442,need for speed iii: hot pursuit,88.0,NaN
639,riven: the sequel to myst,83.0,NaN
647,udraw studio,71.0,NaN
...,...,...,...
16816,space raiders,40.0,NaN
16908,super robot taisen: original generation,72.0,NaN
16935,planet monsters,67.0,NaN
16937,bust-a-move 3000,53.0,NaN


In [38]:
critic_mask['user_score'] = round(critic_mask['critic_score'] / ratio_mean, 1)
critic_mask

,name,critic_score,user_score
182,tomb raider ii,85.0,8.4
302,namco museum: 50th anniversary,61.0,6.0
442,need for speed iii: hot pursuit,88.0,8.7
639,riven: the sequel to myst,83.0,8.2
647,udraw studio,71.0,7.0
...,...,...,...
16816,space raiders,40.0,4.0
16908,super robot taisen: original generation,72.0,7.1
16935,planet monsters,67.0,6.6
16937,bust-a-move 3000,53.0,5.2


In [39]:
user_mask = df.loc[(df['critic_score'].isna()) & (df['user_score'].notna()), ['name', 'critic_score', 'user_score']]
user_mask

,name,critic_score,user_score
255,call of duty: advanced warfare,NaN,4.8
263,call of duty: advanced warfare,NaN,4.9
264,fifa 15,NaN,4.5
292,world soccer winning eleven 9,NaN,8.2
360,just dance 2014,NaN,6.2
...,...,...,...
16875,deal or no deal,NaN,7.5
16882,rugby challenge 3,NaN,6.6
16893,ben 10 omniverse 2,NaN,5.1
16900,bookworm deluxe,NaN,8.0


In [40]:
user_mask['critic_score'] = round(user_mask['user_score'] * ratio_mean, 0)
user_mask

,name,critic_score,user_score
255,call of duty: advanced warfare,49.0,4.8
263,call of duty: advanced warfare,50.0,4.9
264,fifa 15,45.0,4.5
292,world soccer winning eleven 9,83.0,8.2
360,just dance 2014,63.0,6.2
...,...,...,...
16875,deal or no deal,76.0,7.5
16882,rugby challenge 3,67.0,6.6
16893,ben 10 omniverse 2,52.0,5.1
16900,bookworm deluxe,81.0,8.0


Удалось восстановить 1720 значений (1137 + 583). Теперь применим их к основному датафрейму

In [41]:
df.loc[critic_mask.index, 'user_score'] = critic_mask['user_score']
df.loc[user_mask.index, 'critic_score'] = user_mask['critic_score']
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 16954 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16954 non-null  object 
 1   platform         16954 non-null  object 
 2   year_of_release  16954 non-null  float64
 3   genre            16954 non-null  object 
 4   na_sales         16954 non-null  float64
 5   eu_sales         16948 non-null  float64
 6   jp_sales         16950 non-null  float64
 7   other_sales      16954 non-null  float64
 8   critic_score     8825 non-null   float64
 9   user_score       8825 non-null   float64
 10  rating           10085 non-null  object 
dtypes: float64(7), object(4)
memory usage: 2.1+ MB


In [42]:
df.isna().sum()

name                  0
platform              0
year_of_release       0
genre                 0
na_sales              0
eu_sales              6
jp_sales              4
other_sales           0
critic_score       8129
user_score         8129
rating             6869
dtype: int64

<p><code>critic_score</code>, было 8712 пустых, теперь 8129</p>
<p><code>user_score</code>, было 9266 пустых, теперь 8129</p>

<p>Осталось подумать, как бы нам восстановить данные по остальным. Попробуем проверить  зависимость года, жанра и платформы на показатели <code>critic_score</code> и <code>user_score</code>.</p>
<p>Соберем отдельный датафрейм с этими данными.</p>

In [43]:
df_grouped = df.loc[(df['critic_score'].notna()) & (df['user_score'].notna()), ['year_of_release', 'platform', 'genre', 'critic_score', 'user_score']]
df_grouped

,year_of_release,platform,genre,critic_score,user_score
0,2006.0,WII,sports,76.0,8.0
2,2008.0,WII,racing,82.0,8.3
3,2009.0,WII,sports,80.0,8.0
6,2006.0,DS,platform,89.0,8.5
7,2006.0,WII,misc,58.0,6.6
...,...,...,...,...,...
16936,2011.0,PC,shooter,61.0,5.8
16937,2003.0,GC,puzzle,53.0,5.2
16938,2008.0,DS,puzzle,48.0,4.7
16943,2011.0,PC,strategy,60.0,7.2


Можно попробовать сгруппировать данные по средним показателям в разрезе год -> платформа -> жанр и попробовать вычислить среднюю и медиану.

In [44]:
grouped = df_grouped.groupby(['year_of_release', 'platform', 'genre'])[['critic_score', 'user_score']]
summary = grouped.agg(['mean', 'median'])
summary

critic_score        user_score       
                                              mean median       mean median
year_of_release platform genre                                             
1984.0          GBA      platform        75.000000   75.0   8.000000   8.00
1985.0          PC       simulation      59.000000   59.0   5.800000   5.80
1988.0          DS       action          44.000000   44.0   4.400000   4.40
                PC       simulation      64.000000   64.0   2.200000   2.20
                PS3      action          59.000000   59.0   7.700000   7.70
...                                            ...    ...        ...    ...
2016.0          XONE     role-playing    79.333333   83.0   7.733333   7.40
                         shooter         77.000000   80.0   6.575000   6.90
                         simulation      59.000000   59.0   6.350000   6.35
                         sports          79.100000   80.0   6.120000   6.05
                         strategy        87.000000   87.0   8.100000   8.10

[1220 rows x 4 columns]

Попробуем посчитать относительную разницу между средним и медианой

In [45]:
summary['critic_mean_med_ratio'] = ((summary['critic_score']['mean'] - summary['critic_score']['median']) / summary['critic_score']['median']) * 100
summary['user_mean_med_ratio'] = ((summary['user_score']['mean'] - summary['user_score']['median']) / summary['user_score']['median']) * 100
max_critic_diff = summary['critic_mean_med_ratio'].max()
max_user_diff = summary['user_mean_med_ratio'].max()
print(f'Самая большая разница по столбцу critic_score - {max_critic_diff}')
print(f'Самая большая разница по столбцу user_score - {max_user_diff}')

Самая большая разница по столбцу critic_score - 49.33333333333334
Самая большая разница по столбцу user_score - 22.553191489361694


Здесь можно увидеть, что есть большая разница в значениях между mean и median по обоим показателям, поэтому заполним пропуски по медиане. Обработаем сначала вспомогательный датафрейм.

In [46]:
summary = summary[[('critic_score', 'median'), ('user_score', 'median')]]
summary = summary.reset_index()
summary.columns = ['year_of_release', 'platform', 'genre', 'critic_score', 'user_score']
summary

,year_of_release,platform,genre,critic_score,user_score
0,1984.0,GBA,platform,75.0,8.00
1,1985.0,PC,simulation,59.0,5.80
2,1988.0,DS,action,44.0,4.40
3,1988.0,PC,simulation,64.0,2.20
4,1988.0,PS3,action,59.0,7.70
...,...,...,...,...,...
1215,2016.0,XONE,role-playing,83.0,7.40
1216,2016.0,XONE,shooter,80.0,6.90
1217,2016.0,XONE,simulation,59.0,6.35
1218,2016.0,XONE,sports,80.0,6.05


In [47]:
summary.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1220 entries, 0 to 1219
Data columns (total 5 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   year_of_release  1220 non-null   float64
 1   platform         1220 non-null   object 
 2   genre            1220 non-null   object 
 3   critic_score     1220 non-null   float64
 4   user_score       1220 non-null   float64
dtypes: float64(3), object(2)
memory usage: 47.8+ KB


In [48]:
critic_score_result = df.loc[df['critic_score'].isna(), ['platform', 'year_of_release', 'genre', 'critic_score']]
critic_score_result

,platform,year_of_release,genre,critic_score
1,NES,1985.0,platform,NaN
4,GB,1996.0,role-playing,NaN
5,GB,1989.0,puzzle,NaN
9,NES,1984.0,shooter,NaN
10,DS,2005.0,simulation,NaN
...,...,...,...,...
16951,PS3,2016.0,action,NaN
16952,X360,2006.0,sports,NaN
16953,PSV,2016.0,adventure,NaN
16954,GBA,2003.0,platform,NaN


In [49]:
critic_score_result['temp_index'] = critic_score_result.index
critic_score_result = critic_score_result.merge(
    summary[['platform', 'year_of_release', 'genre', 'critic_score']], 
    on=['platform', 'year_of_release', 'genre'], 
    how='left',
    suffixes=('', '_new')
)

critic_score_result.index = critic_score_result['temp_index']
critic_score_result.drop(columns=['temp_index'], inplace=True)

critic_score_result['critic_score'] = critic_score_result['critic_score'].fillna(critic_score_result['critic_score_new'])
critic_score_result.drop(columns=['critic_score_new'], inplace=True)
critic_score_result

,platform,year_of_release,genre,critic_score
temp_index,,,,
1,NES,1985.0,platform,NaN
4,GB,1996.0,role-playing,NaN
5,GB,1989.0,puzzle,NaN
9,NES,1984.0,shooter,NaN
10,DS,2005.0,simulation,70.0
...,...,...,...,...
16951,PS3,2016.0,action,45.0
16952,X360,2006.0,sports,76.5
16953,PSV,2016.0,adventure,78.0


In [50]:
user_score_result = df.loc[df['user_score'].isna(), ['platform', 'year_of_release', 'genre', 'user_score']]
user_score_result

,platform,year_of_release,genre,user_score
1,NES,1985.0,platform,NaN
4,GB,1996.0,role-playing,NaN
5,GB,1989.0,puzzle,NaN
9,NES,1984.0,shooter,NaN
10,DS,2005.0,simulation,NaN
...,...,...,...,...
16951,PS3,2016.0,action,NaN
16952,X360,2006.0,sports,NaN
16953,PSV,2016.0,adventure,NaN
16954,GBA,2003.0,platform,NaN


In [51]:
user_score_result['temp_index'] = user_score_result.index

user_score_result = user_score_result.merge(
    summary[['platform', 'year_of_release', 'genre', 'user_score']], 
    on=['platform', 'year_of_release', 'genre'], 
    how='left',
    suffixes=('', '_new')
)

user_score_result.index = user_score_result['temp_index']
user_score_result.drop(columns=['temp_index'], inplace=True)

user_score_result['user_score'] = user_score_result['user_score'].fillna(user_score_result['user_score_new'])
user_score_result.drop(columns=['user_score_new'], inplace=True)
user_score_result

,platform,year_of_release,genre,user_score
temp_index,,,,
1,NES,1985.0,platform,NaN
4,GB,1996.0,role-playing,NaN
5,GB,1989.0,puzzle,NaN
9,NES,1984.0,shooter,NaN
10,DS,2005.0,simulation,7.80
...,...,...,...,...
16951,PS3,2016.0,action,4.45
16952,X360,2006.0,sports,6.75
16953,PSV,2016.0,adventure,7.80


In [52]:
critic_score_result['user_score'] = user_score_result['user_score']
critic_score_result

,platform,year_of_release,genre,critic_score,user_score
temp_index,,,,,
1,NES,1985.0,platform,NaN,NaN
4,GB,1996.0,role-playing,NaN,NaN
5,GB,1989.0,puzzle,NaN,NaN
9,NES,1984.0,shooter,NaN,NaN
10,DS,2005.0,simulation,70.0,7.80
...,...,...,...,...,...
16951,PS3,2016.0,action,45.0,4.45
16952,X360,2006.0,sports,76.5,6.75
16953,PSV,2016.0,adventure,78.0,7.80


In [53]:
df['critic_score'] = df['critic_score'].fillna(critic_score_result['critic_score'])
df['user_score'] = df['user_score'].fillna(critic_score_result['user_score'])
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 16954 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16954 non-null  object 
 1   platform         16954 non-null  object 
 2   year_of_release  16954 non-null  float64
 3   genre            16954 non-null  object 
 4   na_sales         16954 non-null  float64
 5   eu_sales         16948 non-null  float64
 6   jp_sales         16950 non-null  float64
 7   other_sales      16954 non-null  float64
 8   critic_score     15071 non-null  float64
 9   user_score       15071 non-null  float64
 10  rating           10085 non-null  object 
dtypes: float64(7), object(4)
memory usage: 2.1+ MB


Получилась сложная обработка пропусков, но большинство мы заполнили значениями. Осталось обработать оставшиеся пропуски.

In [54]:
df_empty_scores = df.loc[(df['critic_score'].isna()) & (df['user_score'].isna())]
df_empty_scores = df_empty_scores.loc[(df_empty_scores['year_of_release'] > 1999) & (df_empty_scores['year_of_release'] < 2014)]
df_empty_scores

,name,platform,year_of_release,genre,na_sales,eu_sales,jp_sales,other_sales,critic_score,user_score,rating
59,super mario 64,DS,2004.0,platform,5.01,3.07,1.25,0.97,NaN,NaN,NaN
133,pokémon crystal version,GB,2000.0,role-playing,2.55,1.56,1.29,0.99,NaN,NaN,NaN
402,the legend of zelda: majora's mask,N64,2000.0,action,1.90,0.67,0.73,0.06,NaN,NaN,NaN
551,pokémon stadium 2,N64,2000.0,strategy,1.02,0.36,1.13,0.23,NaN,NaN,NaN
615,perfect dark,N64,2000.0,action,1.55,0.75,0.16,0.06,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...
16876,onigokko! portable,PSP,2013.0,adventure,0.00,0.00,0.01,0.00,NaN,NaN,NaN
16885,time travelers,PSP,2012.0,adventure,0.00,0.00,0.01,0.00,NaN,NaN,NaN
16904,sweet honey coming,PS2,2009.0,adventure,0.00,0.00,0.01,0.00,NaN,NaN,NaN
16922,grisaia no kajitsu: la fruit de la grisaia,PSP,2013.0,adventure,0.00,0.00,0.01,0.00,NaN,NaN,NaN


In [55]:
df_empty_scores.info()

<class 'pandas.core.frame.DataFrame'>
Index: 440 entries, 59 to 16933
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             440 non-null    object 
 1   platform         440 non-null    object 
 2   year_of_release  440 non-null    float64
 3   genre            440 non-null    object 
 4   na_sales         440 non-null    float64
 5   eu_sales         440 non-null    float64
 6   jp_sales         440 non-null    float64
 7   other_sales      440 non-null    float64
 8   critic_score     0 non-null      float64
 9   user_score       0 non-null      float64
 10  rating           42 non-null     object 
dtypes: float64(7), object(4)
memory usage: 41.2+ KB


In [56]:
df_non_empty_scores = df.loc[(df['critic_score'].notna()) & (df['user_score'].notna())]
df_non_empty_scores = df_non_empty_scores.loc[(df_non_empty_scores['year_of_release'] > 1999) & (df_non_empty_scores['year_of_release'] < 2014)]
len(df_empty_scores) / len(df_non_empty_scores) * 100

3.4428794992175273

In [57]:
df_non_empty_scores.info()

<class 'pandas.core.frame.DataFrame'>
Index: 12780 entries, 0 to 16954
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             12780 non-null  object 
 1   platform         12780 non-null  object 
 2   year_of_release  12780 non-null  float64
 3   genre            12780 non-null  object 
 4   na_sales         12780 non-null  float64
 5   eu_sales         12775 non-null  float64
 6   jp_sales         12776 non-null  float64
 7   other_sales      12780 non-null  float64
 8   critic_score     12780 non-null  float64
 9   user_score       12780 non-null  float64
 10  rating           8977 non-null   object 
dtypes: float64(7), object(4)
memory usage: 1.2+ MB


Поскольку по остальным пропускам в critic_score и user_score достаточно сложно восстановить точность данных, а их доля составляет всего 3.5% от нужных нам данных (с 2000 по 2013), думаю можно их в дальнейшем пропустить. Но остался вопрос с рейтингом.

In [58]:
df['genre'].unique()

array(['sports', 'platform', 'racing', 'role-playing', 'puzzle', 'misc',
       'shooter', 'simulation', 'action', 'fighting', 'adventure',
       'strategy'], dtype=object)

Составлю словарь по простой логике жанр -> рейтинг

In [59]:
genre_to_rating = {
    'sports': 'E',
    'platform': 'E',
    'racing': 'E',
    'role-playing': 'T',
    'puzzle': 'E',
    'misc': 'E',
    'shooter': 'M',
    'simulation': 'E',
    'action': 'T',
    'fighting': 'T',
    'adventure': 'E',
    'strategy': 'T'
}

In [60]:
df

,name,platform,year_of_release,genre,na_sales,eu_sales,jp_sales,other_sales,critic_score,user_score,rating
0,wii sports,WII,2006.0,sports,41.36,28.96,3.77,8.45,76.0,8.00,E
1,super mario bros.,NES,1985.0,platform,29.08,3.58,6.81,0.77,NaN,NaN,NaN
2,mario kart wii,WII,2008.0,racing,15.68,12.76,3.79,3.29,82.0,8.30,E
3,wii sports resort,WII,2009.0,sports,15.61,10.93,3.28,2.95,80.0,8.00,E
4,pokemon red/pokemon blue,GB,1996.0,role-playing,11.27,8.89,10.22,1.00,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...
16951,samurai warriors: sanada maru,PS3,2016.0,action,0.00,0.00,0.01,0.00,45.0,4.45,NaN
16952,lma manager 2007,X360,2006.0,sports,0.00,0.01,0.00,0.00,76.5,6.75,NaN
16953,haitaka no psychedelica,PSV,2016.0,adventure,0.00,0.00,0.01,0.00,78.0,7.80,NaN
16954,spirits & spells,GBA,2003.0,platform,0.01,0.00,0.00,0.00,63.0,7.20,NaN


In [61]:
df['rating'] = df['rating'].fillna(df['genre'].map(genre_to_rating))
df

,name,platform,year_of_release,genre,na_sales,eu_sales,jp_sales,other_sales,critic_score,user_score,rating
0,wii sports,WII,2006.0,sports,41.36,28.96,3.77,8.45,76.0,8.00,E
1,super mario bros.,NES,1985.0,platform,29.08,3.58,6.81,0.77,NaN,NaN,E
2,mario kart wii,WII,2008.0,racing,15.68,12.76,3.79,3.29,82.0,8.30,E
3,wii sports resort,WII,2009.0,sports,15.61,10.93,3.28,2.95,80.0,8.00,E
4,pokemon red/pokemon blue,GB,1996.0,role-playing,11.27,8.89,10.22,1.00,NaN,NaN,T
...,...,...,...,...,...,...,...,...,...,...,...
16951,samurai warriors: sanada maru,PS3,2016.0,action,0.00,0.00,0.01,0.00,45.0,4.45,T
16952,lma manager 2007,X360,2006.0,sports,0.00,0.01,0.00,0.00,76.5,6.75,E
16953,haitaka no psychedelica,PSV,2016.0,adventure,0.00,0.00,0.01,0.00,78.0,7.80,E
16954,spirits & spells,GBA,2003.0,platform,0.01,0.00,0.00,0.00,63.0,7.20,E


In [62]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 16954 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16954 non-null  object 
 1   platform         16954 non-null  object 
 2   year_of_release  16954 non-null  float64
 3   genre            16954 non-null  object 
 4   na_sales         16954 non-null  float64
 5   eu_sales         16948 non-null  float64
 6   jp_sales         16950 non-null  float64
 7   other_sales      16954 non-null  float64
 8   critic_score     15071 non-null  float64
 9   user_score       15071 non-null  float64
 10  rating           16954 non-null  object 
dtypes: float64(7), object(4)
memory usage: 2.1+ MB


На этом обработку пропусков пока закончим.

### Явные и неявные дубликаты в данных

<p>Изучим уникальные значения в категориальных данных</p>

Проверим, что работали не с дубликатами по жанрам, платформам, рейтингам и году выпуска.

In [63]:
df['platform'].unique()

array(['WII', 'NES', 'GB', 'DS', 'X360', 'PS3', 'PS2', 'SNES', 'GBA',
       'PS4', '3DS', 'N64', 'PS', 'XB', 'PC', '2600', 'PSP', 'XONE',
       'WIIU', 'GC', 'GEN', 'DC', 'PSV', 'SAT', 'SCD', 'WS', 'NG', 'TG16',
       '3DO', 'GG', 'PCFX'], dtype=object)

In [64]:
df['genre'].unique()

array(['sports', 'platform', 'racing', 'role-playing', 'puzzle', 'misc',
       'shooter', 'simulation', 'action', 'fighting', 'adventure',
       'strategy'], dtype=object)

In [65]:
df['rating'].unique()

array(['E', 'T', 'M', 'E10+', 'K-A', 'AO', 'EC', 'RP'], dtype=object)

In [66]:
df['year_of_release'].unique()

array([2006., 1985., 2008., 2009., 1996., 1989., 1984., 2005., 1999.,
       2007., 2010., 2013., 2004., 1990., 1988., 2002., 2001., 2011.,
       1998., 2015., 2012., 2014., 1992., 1997., 1993., 1994., 1982.,
       2016., 2003., 1986., 2000., 1995., 1991., 1981., 1987., 1980.,
       1983., 1978., 1977., 1979.])

Поскольку мы изначально привели строки к нужным регистрам, вроде работали правильно, без неявных дубликатов.

Проверим наличие явных дубликатов в данных.

In [67]:
df.duplicated().sum()

np.int64(241)

In [68]:
df_cleaned = df.drop_duplicates(subset=None, keep='first', inplace=False)
df_cleaned.info()

<class 'pandas.core.frame.DataFrame'>
Index: 16713 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16713 non-null  object 
 1   platform         16713 non-null  object 
 2   year_of_release  16713 non-null  float64
 3   genre            16713 non-null  object 
 4   na_sales         16713 non-null  float64
 5   eu_sales         16707 non-null  float64
 6   jp_sales         16709 non-null  float64
 7   other_sales      16713 non-null  float64
 8   critic_score     14848 non-null  float64
 9   user_score       14848 non-null  float64
 10  rating           16713 non-null  object 
dtypes: float64(7), object(4)
memory usage: 1.5+ MB


Сделаем промежуточный вывод

Мы нашли 241 явный дубликат - это 100% совпадение строк. Поэтому оставили только 1, а остальные почистили.

Посчитаем количество удаленных строк в абсолютном и относительном значениях.

Мы не зачистили строки с пропусками в user_score и critic_score. Ранее делали вывод, что они не сильно играют роль, поэтому зачистим их сейчас и подведем итоги по общей "Зачистке".

In [69]:
df_cleaned = df_cleaned.loc[(df_cleaned['critic_score'].notna()) & (df_cleaned['user_score'].notna())]
df_cleaned.info()

<class 'pandas.core.frame.DataFrame'>
Index: 14848 entries, 0 to 16954
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             14848 non-null  object 
 1   platform         14848 non-null  object 
 2   year_of_release  14848 non-null  float64
 3   genre            14848 non-null  object 
 4   na_sales         14848 non-null  float64
 5   eu_sales         14842 non-null  float64
 6   jp_sales         14844 non-null  float64
 7   other_sales      14848 non-null  float64
 8   critic_score     14848 non-null  float64
 9   user_score       14848 non-null  float64
 10  rating           14848 non-null  object 
dtypes: float64(7), object(4)
memory usage: 1.4+ MB


У нас было 16956 строк в исходном датафрейме, осталось 14848

In [70]:
16956 - 14848

2108

In [71]:
round((1 - 14848 / 16956) * 100, 1)

12.4

В итоге мы удалили 2108 строку, или 12.4% от исходных данных.

Сделаем промежуточный вывод

<p><b>Промежуточный вывод</b></p>
<p>Мы проделали интересную работу по подготовке данных. Что сделали:</p>
<ol>
    <li>Убрали 2 пустых наименования игр;</li>
    <li>Поменяли наименования столбцов - привели к нижнему регистру и убрали пробелы;</li>
    <li>Привели в порядок типы данных;</li>
    <li>Сразу привели значения в некоторых столбцах к нужным регистрам;</li>
    <li>Полностью восстановили значения годов из открытых источников, их было не много;</li>
    <li>Провели очень сложную работу по восстановлению оценки критиков и пользователей, но вроде получилось;</li>
    <li>Проверили неявные дубликаты и зачистили явные дубликаты;</li>
    <li>Удалили финальные строки с пропущенными оценками критиков и пользователей (3.5% не сильно повлияют на итоговый результат).</li>
</ol>
<p>В итоге мы зачистили 12.4% всех данных и можем приступать к фильтрации.</p>

---

## Фильтрация данных

Коллеги хотят изучить историю продаж игр в начале XXI века, и их интересует период с 2000 по 2013 год включительно. Отберите данные по этому показателю. Сохраните новый срез данных в отдельном датафрейме, например `df_actual`.

In [72]:
df_actual = df_cleaned.query('2000 <= year_of_release <= 2013')
df_actual.info()

<class 'pandas.core.frame.DataFrame'>
Index: 12582 entries, 0 to 16954
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             12582 non-null  object 
 1   platform         12582 non-null  object 
 2   year_of_release  12582 non-null  float64
 3   genre            12582 non-null  object 
 4   na_sales         12582 non-null  float64
 5   eu_sales         12577 non-null  float64
 6   jp_sales         12578 non-null  float64
 7   other_sales      12582 non-null  float64
 8   critic_score     12582 non-null  float64
 9   user_score       12582 non-null  float64
 10  rating           12582 non-null  object 
dtypes: float64(7), object(4)
memory usage: 1.2+ MB


---

## Категоризация данных

Разделим все игры по оценкам пользователей и выделим такие категории: высокая оценка (от 8 до 10 включительно), средняя оценка (от 3 до 8, не включая правую границу интервала) и низкая оценка (от 0 до 3, не включая правую границу интервала).

In [73]:
def categorize_game_user_score(row):
    if row['user_score'] >= 8:
        return "Высокая оценка"
    elif row['user_score'] >= 3:
        return "Средняя оценка"
    else:
        return "Низкая оценка"

df_actual['user_category'] = df_actual.apply(categorize_game_user_score, axis=1)
df_actual

/var/folders/t1/xdgsn1b51y342xx1jlq20cb40000gn/T/ipykernel_3304/1903798311.py:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_actual['user_category'] = df_actual.apply(categorize_game_user_score, axis=1)


,name,platform,year_of_release,genre,na_sales,eu_sales,jp_sales,other_sales,critic_score,user_score,rating,user_category
0,wii sports,WII,2006.0,sports,41.36,28.96,3.77,8.45,76.0,8.00,E,Высокая оценка
2,mario kart wii,WII,2008.0,racing,15.68,12.76,3.79,3.29,82.0,8.30,E,Высокая оценка
3,wii sports resort,WII,2009.0,sports,15.61,10.93,3.28,2.95,80.0,8.00,E,Высокая оценка
6,new super mario bros.,DS,2006.0,platform,11.28,9.14,6.50,2.88,89.0,8.50,E,Высокая оценка
7,wii play,WII,2006.0,misc,13.96,9.18,2.93,2.84,58.0,6.60,E,Средняя оценка
...,...,...,...,...,...,...,...,...,...,...,...,...
16947,men in black ii: alien escape,GC,2003.0,shooter,0.01,0.00,0.00,0.00,70.5,7.50,T,Средняя оценка
16949,woody woodpecker in crazy castle 5,GBA,2002.0,platform,0.01,0.00,0.00,0.00,73.0,7.70,E,Средняя оценка
16950,score international baja 1000: the official game,PS2,2008.0,racing,0.00,0.00,0.00,0.00,61.5,6.75,E,Средняя оценка
16952,lma manager 2007,X360,2006.0,sports,0.00,0.01,0.00,0.00,76.5,6.75,E,Средняя оценка


Разделим все игры по оценкам критиков и выделим такие категории: высокая оценка (от 80 до 100 включительно), средняя оценка (от 30 до 80, не включая правую границу интервала) и низкая оценка (от 0 до 30, не включая правую границу интервала).

In [74]:
def categorize_game_critic_score(row):
    if row['critic_score'] >= 80:
        return "Высокая оценка"
    elif row['critic_score'] >= 30:
        return "Средняя оценка"
    else:
        return "Низкая оценка"

df_actual['critic_category'] = df_actual.apply(categorize_game_critic_score, axis=1)
df_actual

/var/folders/t1/xdgsn1b51y342xx1jlq20cb40000gn/T/ipykernel_3304/2661864744.py:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_actual['critic_category'] = df_actual.apply(categorize_game_critic_score, axis=1)


,name,platform,year_of_release,genre,na_sales,eu_sales,jp_sales,other_sales,critic_score,user_score,rating,user_category,critic_category
0,wii sports,WII,2006.0,sports,41.36,28.96,3.77,8.45,76.0,8.00,E,Высокая оценка,Средняя оценка
2,mario kart wii,WII,2008.0,racing,15.68,12.76,3.79,3.29,82.0,8.30,E,Высокая оценка,Высокая оценка
3,wii sports resort,WII,2009.0,sports,15.61,10.93,3.28,2.95,80.0,8.00,E,Высокая оценка,Высокая оценка
6,new super mario bros.,DS,2006.0,platform,11.28,9.14,6.50,2.88,89.0,8.50,E,Высокая оценка,Высокая оценка
7,wii play,WII,2006.0,misc,13.96,9.18,2.93,2.84,58.0,6.60,E,Средняя оценка,Средняя оценка
...,...,...,...,...,...,...,...,...,...,...,...,...,...
16947,men in black ii: alien escape,GC,2003.0,shooter,0.01,0.00,0.00,0.00,70.5,7.50,T,Средняя оценка,Средняя оценка
16949,woody woodpecker in crazy castle 5,GBA,2002.0,platform,0.01,0.00,0.00,0.00,73.0,7.70,E,Средняя оценка,Средняя оценка
16950,score international baja 1000: the official game,PS2,2008.0,racing,0.00,0.00,0.00,0.00,61.5,6.75,E,Средняя оценка,Средняя оценка
16952,lma manager 2007,X360,2006.0,sports,0.00,0.01,0.00,0.00,76.5,6.75,E,Средняя оценка,Средняя оценка


Проверим результат - сгруппируйем данные по выделенным категориям и посчитаем количество игр в каждой категории.

In [75]:
grouped_user_category = df_actual.groupby('user_category')['name'].count()
grouped_user_category

user_category
Высокая оценка    2944
Низкая оценка      130
Средняя оценка    9508
Name: name, dtype: int64

In [76]:
grouped_critic_category = df_actual.groupby('critic_category')['name'].count()
grouped_critic_category

critic_category
Высокая оценка     2054
Низкая оценка        91
Средняя оценка    10437
Name: name, dtype: int64

Выделим топ-7 платформ по количеству игр, выпущенных за весь актуальный период.

In [77]:
grouped_platform = df_actual.groupby('platform')['name'].count()
grouped_platform

platform
3DS      301
DC        20
DS      2099
GBA      806
GC       548
PC       765
PS       252
PS2     2107
PS3     1106
PS4       15
PSP     1093
PSV      123
WII     1292
WIIU      67
X360    1147
XB       823
XONE      18
Name: name, dtype: int64

In [78]:
grouped_platform.sort_values(ascending=False).head(7)

platform
PS2     2107
DS      2099
WII     1292
X360    1147
PS3     1106
PSP     1093
XB       823
Name: name, dtype: int64

---

## Итоговый вывод

В конце напишите основной вывод и отразите, какую работу проделали. Не забудьте указать описание среза данных и новых полей, которые добавили в исходный датасет.

<p><b>Подготовка данных</b></p>
<p>Мы проделали интересную работу по работе с подготовкой данных. Что сделали перед фильтрацией:</p>
<ol>
    <li>Убрали 2 пустых наименования игр;</li>
    <li>Поменяли наименования столбцов - привели к нижнему регистру и убрали пробелы;</li>
    <li>Привели в порядок типы данных;</li>
    <li>Сразу привели значения в некоторых столбцах к нужным регистрам;</li>
    <li>Полностью восстановили значения годов из открытых источников, их было не много;</li>
    <li>Провели очень сложную работу по восстановлению оценки критиков и пользователей, но вроде получилось;</li>
    <li>Проверили неявные дубликаты и зачистили явные дубликаты;</li>
    <li>Удалили финальные строки с пропущенными оценками критиков и пользователей (3.5% не сильно повлияют на итоговый результат).</li>
</ol>
<p>В итоге мы зачистили 12.4% всех данных.</p>
<p><b>Фильтрация данных</b></p>
<p>Далее мы отфильтровали данные. Что сделали</p>
<ol>
    <li>Создали новый срез <code>df_actual</code> и отобрали нужный нам временной диапазон с 2000 по 2013 год;</li>
    <li>Присвоили нужные категории для оценки критиков и оценки пользователей и добавили новые поля <code>user_category</code> и <code>critic_category</code> в датасет <code>df_actual</code>;</li>
    <li>Посчитали категории по критикам и пользователям;</li>
    <li>Вывели 7 популярных платформ за период с 2000 по 2013 год.</li>
</ol>
<p><b>Выводы</b></p>
<p>Что интересного выянили:</p>
<ol>
    <li>Самой популярной платформой оказалась <code>PS2</code> - 2107 выпущенных игр;</li>
    <li>Критики и игроки в целом были согласны во внениях оценок игр, но игрокам чуть больше понравилимсь выпускаемые игры - 2944 высоких оценок игроков против 2054 критиков.</li>
    <li>Однако, игроки более пристально подходили к оценке качетва игр - 130 низких оценок против 91 у критиков.</li>
</ol>